File: days/day7.swift
#!/usr/bin/env swift
import Foundation


func valid(_ xs: [Int], _ test: Int, ops: Int) -> Bool {
    guard let x = xs.first
    else { return false }
    var acc = [x]
    for y in xs.dropFirst(1) {
        acc = acc.flatMap { x in
            ops == 3 ?
            [
                x + y,
                x * y,
                Int(String(x) + String(y))!,
            ] :
            [
                x + y,
                x * y,
            ]
        }
    }
    return acc.contains(test)
}


func part1(_ data: String, _ ops: Int = 2) -> Int {
    var ans = 0
    for line in data.split(separator: "\n") {
        let cs = line.split(separator: ":")
        let t = Int(cs[0])!
        let xs = cs[1].split(separator: " ").map { Int($0)! }
        if valid(xs, t, ops: ops) {
            ans += t
        }
    }
    return ans
}


func part2(_ data: String) -> Int {
    return part1(data, 3)
}


let test = """
190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20
"""
assert(part1(test) == 3749)
assert(part2(test) == 11387)


let data = try! String(contentsOfFile: "day7.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")

File: days/day7.py
#!/usr/bin/env python
import itertools


def joini(x, y):
    i = y
    while i:
        i //= 10
        x *= 10
    return x + y


ops = [
    int.__add__,
    int.__mul__,
    joini,
]


def valid(xs, t, ops=ops):
    fs = [ops] * (len(xs)-1)
    for ps in itertools.product(*fs):
        x = xs[0]
        for f,y in zip(ps, xs[1:]):
            x = f(x, y)
        if x == t:
            return True


def valid(xs, t, ops=ops):
    rs = [xs[0]]
    for y in xs[1:]:
        rs = [f(x,y) for x in rs for f in ops]
    return t in rs


def part1(data, ops=ops[:2]):
    ans = 0
    for line in data.strip().splitlines():
        n,ns = line.split(':')
        t, *xs = map(int, (n + ns).split())
        if valid(xs, t, ops):
            ans += t
    return ans


def part2(data):
    return part1(data, ops)


data = '''
190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20
'''
assert part1(data) == 3749
assert part2(data) == 11387


data = open('day7.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)
